src/main/java/org/mule/extension/sftp/internal/proxy/AbstractClientProxyConnector.java

```diff
@@ -0,0 +1,191 @@
+/*
+ * Copyright (c) MuleSoft, Inc.  All rights reserved.  http://www.mulesoft.com
+ * The software in this package is published under the terms of the CPAL v1.0
+ * license, a copy of which has been included with this distribution in the
+ * LICENSE.txt file.
+ */
+package org.mule.extension.sftp.internal.proxy;
+
+import org.apache.sshd.client.session.ClientSession;
+import org.apache.sshd.common.io.IoSession;
+import org.apache.sshd.common.util.Readable;
+
+import java.net.InetSocketAddress;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.concurrent.Callable;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicReference;
+
+/**
+ * Basic common functionality for a {@link StatefulProxyConnector}.
+ */
+public abstract class AbstractClientProxyConnector implements StatefulProxyConnector {
+
+  private static final long DEFAULT_PROXY_TIMEOUT_MILLIS = TimeUnit.SECONDS
+      .toMillis(30L);
+
+  /**
+   * Guards {@link #done} and {@link #bufferedCommands}.
+   */
+  private final Object lock = new Object();
```

Comment:
This should be static, otherwise each instance would have a different lock object and it would serve no purpose.